File: Leetcode/Python3/704. Binary Search/solution.py
# https://leetcode.com/problems/binary-search/

from bisect import bisect_left
from typing import List


class Solution:
    def search(self, nums: List[int], target: int) -> int:
        start, end = 0, len(nums) - 1
        while start <= end:
            mid = (start + end) // 2
            if nums[mid] == target:
                return mid
            elif nums[mid] > target:
                end = mid - 1
            else:
                start = mid + 1
        return -1   # When target dose not exist in nums


class Solution:  # Use built-in function
    def search(self, nums: List[int], target: int) -> int:
        index = bisect_left(nums, target)
        if index >= len(nums) or nums[index] != target:
            return -1
        return index

File: Leetcode/Java/704. Binary Search/solution.java
// https://leetcode.com/problems/binary-search/description/

class Solution {
    public int search(int[] nums, int target) {
        int start = 0;
        int end = nums.length - 1;

        while (start <= end) {
            int mid = start + (end - start) / 2;

            if (nums[mid] == target) {
                return mid;
            } else if (nums[mid] > target) {
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        }

        return -1;
    }
}

class Solution { // Recursion
    public int search(int[] nums, int target) {
        return binarySearch(nums, target, 0, nums.length - 1);
    }

    private int binarySearch(int[] nums, int target, int start, int end) {
        if (start <= end) {
            int mid = start + (end - start) / 2;

            if (nums[mid] == target) {
                return mid;
            } else if (nums[mid] > target) {
                return binarySearch(nums, target, start, mid - 1);
            } else {
                return binarySearch(nums, target, mid + 1, end);
            }
        }
        return -1;
    }
}